JDK-8260030: Improve stringStream buffer handling

tstuefe opened the pull request:
stringStream objects are used as temporary string buffers a lot in hotspot. When investigating JDK-8259710, I found that a large majority of the stringStreams created never use the backing buffer fully; about 70% of all streams write less than 32 characters.

stringStream creates an backing buffer, C-heap allocated, with a default size of 256 characters. Some things could be improved:

- add a small in-object buffer for the first n characters to be written. This would avoid or at least delay allocation of the backing buffer from C-heap, at the expense of slightly increased object size and one memcpy when switching over to C-heap. Note that if the backing buffer is smaller than what now is the default size, the total footprint will go down despite the increased object size.

- Delaying allocation of the backing buffer would be good for the many cases where stringStream objects are created as temporary objects without being actually used, see eg. compile.hpp `class PrintInliningBuffer`

- Besides all that, the code could benefit from a little grooming (using modern style syntax etc).

----

This patch:

- renames members of stringStream to conform to common syntax (eg leading underscores) and to be clearer
- Uses initialization lists
- Factors out buffer growth code from stringStream::write() to a new function stringStream::grow()
- Introduces a new object-internal mini buffer, `stringStream::_small_buffer`, sized 32 bytes, which is initially used for all writes

This patch drastically reduces the number of malloc calls done from this class. The internal buffer size of 32byte seems a good cut-off. Running some unrelated test program (no tracing active), I see a reduction in the number of malloc calls from stringStream from ~211K malloc calls down to 53K (debug VM). In a release VM, it drops from ~85K down to about 1K. The reason is that `stringStream` is used in a ton of places as a temporary stack allocated object, to write very minimal text snippets to.

I also tweaked the associated gtest to test more thoroughly.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260030](https://bugs.openjdk.java.net/browse/JDK-8260030): Improve stringStream buffer handling


### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to 4a869bba011a5557aa3e8dfb568e5833c35d3054
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to cced1e7955419ce3c1459e79d0ff0e27e5c355c7


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2160/head:pull/2160`
`$ git checkout pull/2160`

iklam reviewed (approved):
LGTM

tstuefe commented:
> LGTM

Thanks, Ioi!

kimbarrett commented on src/hotspot/share/utilities/ostream.cpp:
```diff
@@ -310,46 +310,66 @@ void outputStream::print_data(void* data, size_t len, bool with_ascii) {
-  size_t end = buffer_pos + len + 1;    // position after write and final '\0'
-  if (end > buffer_length) {
-    if (buffer_fixed) {
+  size_t end = _written + len + 1;    // position after write and final '\0'
```
There's a risk of arithmetic overflow here, if len is "bad".  Can this be
rewritten to be more careful?  Though maybe an assert that end <= _written
is sufficient, as line 369 already protects against buffer overrun.
Also, the comments aren't lined up.

tstuefe replied:
I'll rewrite this coding to be safer.

kimbarrett commented on src/hotspot/share/utilities/ostream.cpp:
```diff
@@ -384,8 +404,8 @@ char* stringStream::as_string(bool c_heap) const {
 stringStream::~stringStream() {
-  if (buffer_fixed == false && buffer != NULL) {
-    FREE_C_HEAP_ARRAY(char, buffer);
+  if (_is_fixed == false && _buffer != NULL && _buffer != _small_buffer) {
```
`_is_fixed == false` => `!_is_fixed`
Is `_buffer == NULL` actually possible now?

tstuefe replied:
No, you are right. I'll remove it. free would accept NULL anyway, so there is not even a risk.

kimbarrett commented on src/hotspot/share/utilities/ostream.hpp:
```diff
@@ -193,27 +193,31 @@ class ttyUnlocker: StackObj {
+  size_t _written;
+  size_t _capacity;
+  const bool _is_fixed;
+  char   _small_buffer[32];
```
I'm a little surprised that 32 bytes is that effective. Any idea how much better
(in malloc avoidance) a few more words would be?  Since 32 bytes isn't really
all that much.

tstuefe replied:
I was surprised too. I did not want to blow up the stringStream object too much because I was afraid someone may use it on stack in a recursion. I even considered melting the buffer and the buffer pointer into a union to save space, since you only need either one or the other, but the coding got too complex and it felt overengineered.

I did some tests:

java -version

```
debug:    <32:1038 (86%), <48:111 (9%), <64:42 (3%), >=64:4 (0%)
release:  <32:449 (84%),  <48:54 (10%), <64:26 (4%), >=64:4 (0%)
```

Running some unrelated test program, no tracing, doing metaspace stuff

```
debug: 	  <32:151985 (52%), <48:83341 (28%), <64:46916 (16%), >=64:6445 (2%)                  
release:  <32:84123 (49%),  <48:80332 (47%), <64:871 (0%), >=64:3567 (2%) 
```

same test with -XX:+PrintOptoInlining, which creates many temporary stringStream objects with large buffers

```
debug:	  <32:160311 (52%), <48:83661 (27%), <64:47821 (15%), >=64:15345 (4%)   
```

So, depending on the scenario, increasing the buffer size to 48 may give us between 10% and almost 30% more cases. Beyond that the effect strongly diminishes. I will increase it to 48.

kimbarrett commented on src/hotspot/share/utilities/ostream.cpp:
```diff
@@ -360,21 +380,21 @@ void stringStream::write(const char* s, size_t len) {
-  strncpy(copy, buffer, buffer_pos);
-  copy[buffer_pos] = 0;  // terminating null
+    NEW_C_HEAP_ARRAY(char, _written + 1, mtInternal) : NEW_RESOURCE_ARRAY(char, _written + 1);
+  strncpy(copy, _buffer, _written);
```
Pre-existing: Should this really be using strncpy rather than memcpy?
strncpy will stop at an embedded NUL.

tstuefe replied:
Sure, I'll switch it to memcpy. Its probably more efficient. The content should not contain embedded zeros, but since we do not test on write, they can slip in. Using memcpy would be at least consistent with write().

tstuefe commented on src/hotspot/share/utilities/ostream.cpp:
```diff
@@ -310,46 +310,66 @@ void outputStream::print_data(void* data, size_t len, bool with_ascii) {
+void stringStream::grow(size_t new_capacity) {
+  assert(new_capacity > _capacity, "Sanity");
+  assert(new_capacity > sizeof(_small_buffer), "Sanity");
+  assert(!_is_fixed, "Don't call for caller provided buffers");
```
Will do.

tstuefe commented on src/hotspot/share/utilities/ostream.hpp:
```diff
@@ -193,27 +193,31 @@ class ttyUnlocker: StackObj {
-  size_t buffer_pos;
-  size_t buffer_length;
-  bool   buffer_fixed;
+  char*  _buffer;
```
Very good, I'll do it.

tstuefe commented on src/hotspot/share/utilities/ostream.cpp:
```diff
@@ -360,21 +380,21 @@ void stringStream::write(const char* s, size_t len) {
 
 void stringStream::reset() {
-  buffer_pos = 0; _precount = 0; _position = 0;
+  _written = 0; _precount = 0; _position = 0;
```
Strictly speaking yes. But _newlines is really only used as a very circumvent way to trigger a flush in defaultStream::write() if the last write contained a newline. We could probably throw _newlines completely away at the cost of a `strchr(s, '\n')` in defaultStream::write(). I'll remove the outside accessor `newlines()` at least, since its not used, and will correct stringStream::reset(), so it is consistent for now. I leave the cleanup for another day.

tstuefe commented:
I rewrote stringStream::write somewhat:
- I check and ignore len==0
- I check len for unreasonably high values (>1G). I assert, in release I return: stringStream is used for tracing a lot and we never touch all tracing code in debug. I did not add a overflow check, though I could, since overflow would require very high values for _written (> SIZE_MAX - 1G) which is not possible.
- I changed the guarantee to an assert, and added a similar assert to the start of the function

kimbarrett commented on src/hotspot/share/utilities/ostream.cpp:
```diff
@@ -404,7 +411,7 @@ char* stringStream::as_string(bool c_heap) const {
 
 stringStream::~stringStream() {
-  if (_is_fixed == false && _buffer != NULL && _buffer != _small_buffer) {
+  if (_is_fixed == false && _buffer != _small_buffer) {
```
Still using `_is_fixed == false` rather than `!_is_fixed`.

tstuefe replied:
Oh, I missed that, sorry. Will fix before integration.

kimbarrett reviewed (approved):
Other than the minor nit about use of `_is_fixed` in the destructor, looks good.

tstuefe commented:
Thanks Kim, Ioi.

/integrate
